test/e2e/upgrade/upgrade_test.go

```diff
@@ -0,0 +1,185 @@
+package upgrade
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"strings"
+	"testing"
+
+	"github.com/iancoleman/strcase"
+	ispnv1 "github.com/infinispan/infinispan-operator/api/v1"
+	"github.com/infinispan/infinispan-operator/api/v2alpha1"
+	"github.com/infinispan/infinispan-operator/controllers"
+	"github.com/infinispan/infinispan-operator/controllers/constants"
+	batchtest "github.com/infinispan/infinispan-operator/test/e2e/batch"
+	tutils "github.com/infinispan/infinispan-operator/test/e2e/utils"
+	"github.com/operator-framework/api/pkg/manifests"
+	coreos "github.com/operator-framework/api/pkg/operators/v1alpha1"
+	"github.com/stretchr/testify/assert"
+	"gopkg.in/yaml.v2"
+	corev1 "k8s.io/api/core/v1"
+	apiextv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+var (
+	ctx                   = context.TODO()
+	testKube              = tutils.NewTestKubernetes(os.Getenv("TESTING_CONTEXT"))
+	subName               = constants.GetEnvWithDefault("SUBSCRIPTION_NAME", "infinispan-operator")
+	subNamespace          = constants.GetEnvWithDefault("SUBSCRIPTION_NAMESPACE", tutils.Namespace)
+	catalogSource         = constants.GetEnvWithDefault("SUBSCRIPTION_CATALOG_SOURCE", "test-catalog")
+	catalogSourcNamespace = constants.GetEnvWithDefault("SUBSCRIPTION_CATALOG_SOURCE_NAMESPACE", tutils.Namespace)
+	subPackage            = constants.GetEnvWithDefault("SUBSCRIPTION_PACKAGE", "infinispan")
+
+	packageManifest = testKube.PackageManifest(subPackage, catalogSourcNamespace)
+	sourceChannel   = getChannel("SUBSCRIPTION_CHANNEL_SOURCE", 1, packageManifest)
+	targetChannel   = getChannel("SUBSCRIPTION_CHANNEL_TARGET", 0, packageManifest)
+
+	subStartingCsv = constants.GetEnvWithDefault("SUBSCRIPTION_STARTING_CSV", sourceChannel.CurrentCSVName)
+)
+
+func TestUpgrade(t *testing.T) {
+	printManifest()
+	testKube.NewNamespace(tutils.Namespace)
+
+	replicas := 2
+	sub := &coreos.Subscription{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: coreos.SubscriptionCRDAPIVersion,
+			Kind:       coreos.SubscriptionKind,
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      subName,
+			Namespace: subNamespace,
+		},
+		Spec: &coreos.SubscriptionSpec{
+			Channel:                sourceChannel.Name,
+			CatalogSource:          catalogSource,
+			CatalogSourceNamespace: catalogSourcNamespace,
+			InstallPlanApproval:    coreos.ApprovalManual,
+			Package:                subPackage,
+			StartingCSV:            subStartingCsv,
+		},
+	}
+	testKube.CreateSubscription(sub)
+	defer testKube.DeleteNamespace(tutils.Namespace)
```

Comment:
We should probably also add defer on removing subscription and csv in case `SUBSCRIPTION_NAMESPACE` != tutils.Namespace

Reply:
I've added a deferred call to `cleanup(..)` that will remove the created OLM CRs as well as any Infinispan resources.